### PR TITLE
Add DecodeDN

### PR DESCRIPTION
**Pull request by taran-p**
Adds DecodeDN to ldap package

### CODE REVIEW

File: ldap/decode.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
+// The MIT License (MIT)
```
Comment (harshavardhana) [inline ldap/decode.go:1]: filename must be `decode_dn_contrib.go`

File: ldap/decode_test.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2015-2022 MinIO, Inc.
```
Comment (harshavardhana) [inline ldap/decode_test.go:1]:
```suggestion
// Copyright (c) 2015-2024 MinIO, Inc.
```

File: ldap/decode_test.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+			input:    `cn=pr\c3\bcfen,dc=example,dc=com`,
+			expected: "cn=prüfen,dc=example,dc=com",
+		},
+		{
+			input: `cn=foo,dc=example,dc=com\`,
+			err:   fmt.Errorf("got corrupted escaped character: '%s'", `cn=foo,dc=example,dc=com\`),
+		},
+		{
+			input: `cn=foo,dc=example,dc=com\a`,
+			err:   fmt.Errorf("ailed to decode escaped character: encoding/hex: invalid byte: %s", "a"),
```
Comment (harshavardhana) [inline ldap/decode_test.go:54]: please 

- add one test for `chinese` character
- one for Umlauts
- one for unicode '.'
- one for Cyrillic
- one with Devanagari (if supported not sure, please check)
- any other esoteric characters that you can think of

Some are added, and some are not just laying out what we expect to cover all cases..

Comment (harshavardhana) [inline ldap/decode_test.go:54]:
```suggestion
		{
			input:    `cn=\d0\98\d0\b2\d0\b0\d0\bd\d0\be\d0\b2 \d0\98\d0\b2\d0\b0\d0\bd,dc=example,dc=com`,
			expected: "cn=Иванов Иван,dc=example,dc=com",
		},
		{
			input:    `cn=\20foo,dc=example,dc=com`,
			expected: "cn= foo,dc=example,dc=com",
		},
		{
			input:    `cn=pr\c3\bcfen,dc=example,dc=com`,
			expected: "cn=prüfen,dc=example,dc=com",
		},
		{
			input: `cn=foo,dc=example,dc=com\`,
			err:   fmt.Errorf("got corrupted escaped character: '%s'", `cn=foo,dc=example,dc=com\`),
		},
		{
			input: `cn=foo,dc=example,dc=com\a`,
			err:   fmt.Errorf("unable to decode escaped character: encoding/hex: invalid byte: %s", "a"),
```

File: ldap/decode.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+			builder.WriteRune(s[i+1])
+			i++
+			continue
+		}
+
+		// If the escaped character is not a special character, it should
+		// be a hex-encoded character of the form \XX if it's not at least
+		// two characters long, it's a corrupted escaped character
+		if i+2 >= len(s) {
+			return "", errors.New("failed to decode escaped character: encoding/hex: invalid byte: " + string(s[i+1]))
```
Comment (harshavardhana) [inline ldap/decode.go:71]:
```suggestion
			return "", errors.New("unable to decode escaped character: encoding/hex: invalid byte: " + string(s[i+1]))
```

File: ldap/decode.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+		// Get the runes for the two characters after the escape character
+		// and convert them to a byte slice
+		xx := []byte(string(s[i+1 : i+3]))
+
+		// If the two runes are not hex characters and result in more than
+		// two bytes when converted to a byte slice, it's a corrupted
+		// escaped character
+		if len(xx) != 2 {
+			return "", errors.New("failed to decode escaped character: invalid byte: " + string(xx))
```
Comment (harshavardhana) [inline ldap/decode.go:82]:
```suggestion
			return "", errors.New("unable to decode escaped character: invalid byte: " + string(xx))
```

File: ldap/decode.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+		}
+
+		// Decode the hex-encoded character and add it to the builder
+		dst := []byte{0}
+		if n, err := hex.Decode(dst, xx); err != nil {
+			return "", errors.New("failed to decode escaped character: " + err.Error())
```
Comment (harshavardhana) [inline ldap/decode.go:88]:
```suggestion
			return "", errors.New("unable to decode escaped character: " + err.Error())
```

File: ldap/decode.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+		} else if n != 1 {
+			return "", fmt.Errorf("failed to decode escaped character: encoding/hex: expected 1 byte when un-escaping, got %d", n)
```
Comment (harshavardhana) [inline ldap/decode.go:90]:
```suggestion
			return "", fmt.Errorf("unable to decode escaped character: encoding/hex: expected 1 byte when un-escaping, got %d", n)
```

File: ldap/decode.go (hunk 5 above)
Comment (harshavardhana) [inline ldap/decode.go:82]: lets use 

```
return "", ldap.NewError(34, fmt.Errorf("unable to decode escaped character: invalid byte: %s", string(xx))
```

Then use this on MinIO server to send a proper error response back to client as a 400 Bad Request with "message description" with this error.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: ldap/decode.go (hunk 5 above)
Comment (taran-p) [reply ldap/decode.go:82]: For this specific error or all the returned errors?

Comment (harshavardhana) [reply ldap/decode.go:82]: all of them this was just an example, this is to ensure that we do not return back 500 Internal Server error instead have a proper error propagation in-place.

**Comment by taran-p** [comment]
All changes should be there

**Review by donatello — APPROVED** [review]
LGTM - though we should rename `ldap/decode_test.go` to `ldap/decode_dn_contrib_test.go

**Comment by taran-p** [comment]
Forgot about that, fixed

**Review by harshavardhana — APPROVED** [review]

File: ldap/decode_dn_contrib_test.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package ldap
+
+import (
+	"errors"
+	"fmt"
+	"testing"
+)
+
```
Comment (harshavardhana) [inline ldap/decode_dn_contrib_test.go:25]: No this file doesn't need to renamed() this test is entire written by us so its under our copyright not a _contrib @taran-p @donatello

**Pull request merged**